File: app/presentation/pages/history_page.py
"""History page: recent compression jobs with stats."""

from __future__ import annotations

from pathlib import Path

from PySide6.QtCore import Qt
from PySide6.QtWidgets import QFrame, QHBoxLayout, QScrollArea, QVBoxLayout, QWidget

from app.core.entities import HistoryEntry
from app.presentation.app_controller import AppController
from app.presentation.components.badge import Badge, BadgeVariant
from app.presentation.components.button import Button, ButtonSize, ButtonVariant
from app.presentation.components.card import Card
from app.presentation.components.filetable import Column, FileTable
from app.presentation.components.modal import confirm
from app.presentation.components.typography import SectionTitle, SecondaryText
from app.presentation.components.utils import format_size, format_timestamp
from app.presentation.theme.registry import active_theme


class HistoryPage(QWidget):
    def __init__(self, controller: AppController, parent=None) -> None:
        super().__init__(parent)
        self.controller = controller
        self._entries: list[HistoryEntry] = []
        self._build()
        self.refresh()

    # ------------------------------------------------------------------ #
    def _build(self) -> None:
        outer = QVBoxLayout(self)
        outer.setContentsMargins(0, 0, 0, 0)

        scroll = QScrollArea()
        scroll.setWidgetResizable(True)
        scroll.setHorizontalScrollBarPolicy(Qt.ScrollBarAlwaysOff)
        outer.addWidget(scroll)

        container = QWidget()
        container.setObjectName("pageRoot")
        layout = QVBoxLayout(container)
        layout.setContentsMargins(28, 22, 28, 28)
        layout.setSpacing(16)

        # header
        header = QHBoxLayout()
        title_col = QVBoxLayout()
        title_col.setSpacing(2)
        title_col.addWidget(SectionTitle("History"))
        title_col.addWidget(SecondaryText("Everything you have compressed with Compresstor."))
        header.addLayout(title_col)
        header.addStretch(1)
        self._clear_btn = Button("Clear history", ButtonVariant.GHOST, ButtonSize.SM, icon="trash-2", icon_size=14)
        self._clear_btn.clicked.connect(self._on_clear)
        header.addWidget(self._clear_btn)
        layout.addLayout(header)

        # stats strip
        stats_card = Card()
        stats_layout = stats_card.body()
        stats_layout.setSpacing(0)
        stats_row = QHBoxLayout()
        stats_row.setSpacing(24)
        self._stat_files = self._stat_block(stats_row, "files", "0")
        self._stat_saved = self._stat_block(stats_row, "saved", "0 MB")
        self._stat_avg = self._stat_block(stats_row, "avg savings", "0%")
        self._stat_today = self._stat_block(stats_row, "today", "0")
        stats_row.addStretch(1)
        stats_layout.addLayout(stats_row)
        layout.addWidget(stats_card)

        # table card
        table_card = Card("Recent Activity")
        table_layout = table_card.body()
        self._table = FileTable([
            Column("file", "File", 0),
            Column("kind", "Type", 90),
            Column("before", "Before", 110, align="right"),
            Column("after", "After", 110, align="right"),
            Column("savings", "Saved", 120),
            Column("date", "Date", 170),
            Column("actions", "", 44, sortable=False, align="center"),
        ])
        self._table.set_filter_options([("pdf", "PDF"), ("image", "Image")])
        self._table.context_menu(self._on_menu)
        table_layout.addWidget(self._table)
        layout.addWidget(table_card, 1)

        scroll.setWidget(container)

    def _stat_block(self, row: QHBoxLayout, label: str, value: str) -> "QLabel":
        from app.presentation.components.typography import CaptionText

        block = QVBoxLayout()
        block.setSpacing(2)
        caption = CaptionText(label.upper())
        block.addWidget(caption)
        from app.presentation.components.typography import NormalText

        value_label = NormalText(value)
        value_label.setStyleSheet("font-size: 18pt; font-weight: 700; background: transparent; color: palette(text);")
        block.addWidget(value_label)
        row.addLayout(block)
        return value_label

    # ------------------------------------------------------------------ #
    def refresh(self) -> None:
        self._entries = self.controller.history_uc.list()
        rows = [self._row_for(e) for e in self._entries]
        self._table.set_rows(rows)
        self._update_stats()

    def _update_stats(self) -> None:
        done = [e for e in self._entries if e.status == "done"]
        total_saved = sum(max(0, e.original_size - e.compressed_size) for e in done)
        avg = (sum(e.savings_percent for e in done) / len(done)) if done else 0.0
        from datetime import datetime

        today = sum(1 for e in self._entries if datetime.fromtimestamp(e.timestamp).date() == datetime.now().date())
        self._stat_files.setText(str(len(done)))
        self._stat_saved.setText(format_size(total_saved))
        self._stat_avg.setText(f"{avg:.0f}%")
        self._stat_today.setText(str(today))

    def _row_for(self, entry: HistoryEntry) -> dict:
        icon = "file-text" if entry.kind == "pdf" else "image"
        kind_variant = BadgeVariant.DEFAULT if entry.kind == "pdf" else BadgeVariant.INFO
        status_variant = BadgeVariant.SUCCESS if entry.status == "done" else (
            BadgeVariant.DANGER if entry.status == "failed" else BadgeVariant.WARNING)
        return {
            "file": {"kind": "text", "text": entry.file_name, "secondary": entry.output_path, "icon": icon, "sort": entry.file_name.lower()},
            "kind": {"kind": "badge", "text": "PDF" if entry.kind == "pdf" else "Image", "variant": kind_variant},
            "before": {"kind": "text", "text": format_size(entry.original_size), "sort": entry.original_size},
            "after": {"kind": "text", "text": format_size(entry.compressed_size), "sort": entry.compressed_size},
            "savings": {"kind": "badge", "text": f"{entry.savings_percent:.0f}% smaller", "variant": status_variant,
                        "icon": "trending-down" if entry.savings_percent > 0 else None},
            "date": {"kind": "text", "text": format_timestamp(entry.timestamp), "sort": entry.timestamp},
            "actions": {"kind": "button", "icon": "more-horizontal", "tooltip": "Actions", "callback": lambda e=entry: self._entry_actions(e)},
        }

    # ------------------------------------------------------------------ #
    def _on_menu(self, selected: list[int], menu) -> None:
        if not selected:
            return
        entry = self._entries[selected[0]]
        menu.addAction(active_theme().icon("folder-open", size=15), "Open file location",
                       lambda: self._open_location(entry.output_path))
        menu.addAction(active_theme().icon("copy", size=15), "Copy output path",
                       lambda: self._copy(entry.output_path))
        menu.addSeparator()
        menu.addAction(active_theme().icon("trash-2", size=15), "Remove from history",
                       lambda: self._remove_entry(entry))

    def _entry_actions(self, entry: HistoryEntry) -> None:
        from PySide6.QtWidgets import QApplication, QMenu

        menu = QMenu(self)
        menu.addAction(active_theme().icon("folder-open", size=15), "Open file location",
                       lambda: self._open_location(entry.output_path))
        menu.addAction(active_theme().icon("copy", size=15), "Copy output path",
                       lambda: self._copy(entry.output_path))
        menu.addSeparator()
        menu.addAction(active_theme().icon("trash-2", size=15), "Remove from history",
                       lambda: self._remove_entry(entry))
        button = self.sender()
        if button:
            menu.exec(button.mapToGlobal(button.rect().bottomLeft()))

    def _remove_entry(self, entry: HistoryEntry) -> None:
        self.controller.history_uc.remove(entry)
        self.refresh()

    def _open_location(self, path: str) -> None:
        import subprocess
        import sys

        p = Path(path)
        if not p.exists():
            p = Path(path).parent if Path(path).parent.exists() else Path.home()
        if sys.platform == "darwin":
            subprocess.Popen(["open", "-R", str(p)])
        elif sys.platform == "win32":
            subprocess.Popen(["explorer", "/select,", str(p)])
        else:
            subprocess.Popen(["xdg-open", str(p.parent)])

    def _copy(self, text: str) -> None:
        from PySide6.QtWidgets import QApplication

        QApplication.clipboard().setText(text)

    def _on_clear(self) -> None:
        if not self._entries:
            return
        confirm(
            self.window(),
            "Clear history?",
            "All compression history will be removed. This cannot be undone.",
            confirm_text="Clear history",
            danger=True,
            on_result=lambda ok: self._clear_now() if ok else None,
        )

    def _clear_now(self) -> None:
        self.controller.history_uc.clear()
        self.refresh()

File: flutter/lib/pages/history_page.dart
// History — full CRUD page (Phase 4). Lists compression history from the
// engine, shows aggregate stats, per-row actions (open file, copy path,
// remove), and a "Clear history" button with confirmation.

import 'dart:io';

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

import '../components/badge.dart';
import '../components/button.dart';
import '../components/card.dart';
import '../components/modal.dart';
import '../components/toast.dart';
import '../engine/format.dart';
import '../state/app_controller.dart';
import '../state/app_scope.dart';
import '../theme/app_theme.dart';
import '../theme/icons.dart';
import '../theme/spacing.dart';

class HistoryPage extends StatefulWidget {
  const HistoryPage({super.key});

  @override
  State<HistoryPage> createState() => _HistoryPageState();
}

class _HistoryPageState extends State<HistoryPage> {
  bool _loadScheduled = false;

  AppController get _c => AppScope.of(context);

  void _scheduleLoad() {
    if (_loadScheduled) return;
    final c = _c;
    if (c.historyEntries.isEmpty && !c.isLoadingHistory) {
      _loadScheduled = true;
      WidgetsBinding.instance.addPostFrameCallback((_) {
        if (mounted) c.loadHistory();
      });
    }
  }

  // ---------------------------------------------------------------- stats --

  List<Map<String, dynamic>> get _done =>
      _c.historyEntries.where((e) => e['status'] == 'done').toList();

  int get _totalSaved => _done.fold(0, (s, e) {
    final orig = (e['original_size'] as num?)?.toInt() ?? 0;
    final comp = (e['compressed_size'] as num?)?.toInt() ?? 0;
    return s + (orig - comp).clamp(0, orig);
  });

  double get _avgSavings {
    final d = _done;
    if (d.isEmpty) return 0;
    return d.fold(0.0, (s, e) {
      final pct = (e['savings_percent'] as num?)?.toDouble() ?? 0;
      return s + pct;
    }) / d.length;
  }

  int get _todayCount => _c.historyEntries
      .where((e) => isToday((e['timestamp'] as num?)?.toDouble() ?? 0))
      .length;

  // --------------------------------------------------------------- build --

  @override
  Widget build(BuildContext context) {
    _scheduleLoad();
    final theme = AppTheme.of(context);
    final palette = theme.palette;
    final entries = _c.historyEntries;
    final loading = _c.isLoadingHistory;

    return SingleChildScrollView(
      padding: const EdgeInsets.only(left: 28, right: 28, top: 22, bottom: 28),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          _buildHeader(theme),
          const SizedBox(height: 16),
          _buildStats(theme),
          const SizedBox(height: 16),
          AppCard(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.stretch,
              children: [
                Row(
                  children: [
                    Text('Recent Activity', style: theme.typography.cardTitle),
                    const SizedBox(width: 8),
                    AppBadge(
                      label: '${entries.length}',
                      tone: BadgeTone.neutral,
                    ),
                  ],
                ),
                const SizedBox(height: 12),
                if (loading && entries.isEmpty)
                  Padding(
                    padding: const EdgeInsets.all(24),
                    child: Center(
                      child: Text('Loading…', style: theme.typography.secondary),
                    ),
                  )
                else if (entries.isEmpty)
                  Padding(
                    padding: const EdgeInsets.all(24),
                    child: Center(
                      child: Text(
                        'No history yet. Compress some files to see them here.',
                        style: theme.typography.secondary,
                      ),
                    ),
                  )
                else ...[
                  _tableHeader(theme, palette),
                  for (final e in entries) _tableRow(theme, palette, e),
                ],
              ],
            ),
          ),
        ],
      ),
    );
  }

  // -------------------------------------------------------------- header --

  Widget _buildHeader(AppTheme theme) {
    return Row(
      crossAxisAlignment: CrossAxisAlignment.end,
      children: [
        Expanded(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Text('History', style: theme.typography.pageTitle),
              const SizedBox(height: 2),
              Text(
                'Everything you have compressed with Compresstor.',
                style: theme.typography.secondary,
              ),
            ],
          ),
        ),
        AppButton(
          label: 'Refresh',
          icon: 'refresh-cw',
          variant: ButtonVariant.outline,
          size: ButtonSize.sm,
          onPressed: _c.isLoadingHistory ? null : () => _c.loadHistory(),
        ),
        const SizedBox(width: 8),
        AppButton(
          label: 'Clear history',
          icon: 'trash-2',
          variant: ButtonVariant.ghost,
          size: ButtonSize.sm,
          onPressed: _c.historyEntries.isEmpty ? null : _onClear,
        ),
      ],
    );
  }

  // --------------------------------------------------------------- stats --

  Widget _buildStats(AppTheme theme) {
    return AppCard(
      child: Row(
        children: [
          _statBlock(theme, 'FILES', '${_done.length}'),
          const SizedBox(width: 24),
          _statBlock(theme, 'SAVED', formatSize(_totalSaved)),
          const SizedBox(width: 24),
          _statBlock(theme, 'AVG SAVINGS', '${_avgSavings.round()}%'),
          const SizedBox(width: 24),
          _statBlock(theme, 'TODAY', '$_todayCount'),
          const Spacer(),
        ],
      ),
    );
  }

  Widget _statBlock(AppTheme theme, String label, String value) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text(label, style: theme.typography.caption),
        const SizedBox(height: 2),
        Text(
          value,
          style: theme.typography.pageTitle.copyWith(fontSize: 22),
        ),
      ],
    );
  }

  // --------------------------------------------------------------- table --

  Widget _tableHeader(AppTheme theme, palette) {
    final style = theme.typography.caption;
    return Container(
      padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 8),
      decoration: BoxDecoration(
        border: Border(bottom: BorderSide(color: palette.border)),
      ),
      child: Row(
        children: [
          Expanded(flex: 4, child: Text('File', style: style)),
          Expanded(flex: 2, child: Text('Type', style: style)),
          Expanded(
            flex: 2,
            child: Text('Before', style: style, textAlign: TextAlign.right),
          ),
          Expanded(
            flex: 2,
            child: Text('After', style: style, textAlign: TextAlign.right),
          ),
          Expanded(flex: 2, child: Text('Saved', style: style)),
          Expanded(flex: 2, child: Text('Date', style: style)),
          const SizedBox(width: 44),
        ],
      ),
    );
  }

  Widget _tableRow(
    AppTheme theme,
    dynamic palette,
    Map<String, dynamic> entry,
  ) {
    final name = entry['file_name'] as String? ?? '';
    final kind = entry['kind'] as String? ?? '';
    final origSize = (entry['original_size'] as num?)?.toInt() ?? 0;
    final compSize = (entry['compressed_size'] as num?)?.toInt() ?? 0;
    final pct = (entry['savings_percent'] as num?)?.toDouble() ?? 0;
    final status = entry['status'] as String? ?? '';
    final ts = (entry['timestamp'] as num?)?.toDouble() ?? 0;
    final outputPath = entry['output_path'] as String? ?? '';

    final kindTone = kind == 'pdf' ? BadgeTone.neutral : BadgeTone.info;
    final savingsTone = status == 'done'
        ? BadgeTone.success
        : status == 'failed'
            ? BadgeTone.danger
            : BadgeTone.warning;

    return Container(
      padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 10),
      decoration: BoxDecoration(
        border: Border(bottom: BorderSide(color: palette.borderSoft as Color)),
      ),
      child: Row(
        children: [
          Expanded(
            flex: 4,
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Row(
                  children: [
                    AppIcon(
                      kind == 'pdf' ? 'file-text' : 'image',
                      size: 14,
                      color: palette.textSecondary as Color,
                    ),
                    const SizedBox(width: 6),
                    Expanded(
                      child: Text(
                        name,
                        style: theme.typography.body,
                        overflow: TextOverflow.ellipsis,
                      ),
                    ),
                  ],
                ),
                if (outputPath.isNotEmpty)
                  Padding(
                    padding: const EdgeInsets.only(left: 20),
                    child: Text(
                      outputPath,
                      style: theme.typography.caption,
                      overflow: TextOverflow.ellipsis,
                    ),
                  ),
              ],
            ),
          ),
          Expanded(
            flex: 2,
            child: AppBadge(
              label: kind == 'pdf' ? 'PDF' : 'Image',
              tone: kindTone,
            ),
          ),
          Expanded(
            flex: 2,
            child: Text(
              formatSize(origSize),
              style: theme.typography.secondary,
              textAlign: TextAlign.right,
            ),
          ),
          Expanded(
            flex: 2,
            child: Text(
              formatSize(compSize),
              style: theme.typography.secondary,
              textAlign: TextAlign.right,
            ),
          ),
          Expanded(
            flex: 2,
            child: AppBadge(
              label: '${pct.round()}% smaller',
              tone: savingsTone,
            ),
          ),
          Expanded(
            flex: 2,
            child: Text(
              formatTimestamp(ts),
              style: theme.typography.secondary,
            ),
          ),
          SizedBox(
            width: 44,
            child: _ActionsMenu(
              onOpenLocation: () => _openLocation(outputPath),
              onCopyPath: () => _copyPath(outputPath),
              onRemove: () => _removeEntry(entry),
            ),
          ),
        ],
      ),
    );
  }

  // ------------------------------------------------------------- actions --

  void _openLocation(String path) {
    final target = File(path).existsSync()
        ? path
        : File(path).parent.existsSync()
            ? File(path).parent.path
            : Platform.environment['HOME'] ?? '/';
    if (Platform.isMacOS) {
      Process.run('open', ['-R', target]);
    } else if (Platform.isWindows) {
      Process.run('explorer', ['/select,', target]);
    } else {
      Process.run('xdg-open', [File(target).parent.path]);
    }
  }

  void _copyPath(String path) {
    Clipboard.setData(ClipboardData(text: path));
    if (mounted) {
      ToastHost.of(context).info('Copied', 'Output path copied to clipboard.');
    }
  }

  void _removeEntry(Map<String, dynamic> entry) {
    final ts = (entry['timestamp'] as num?)?.toDouble() ?? 0;
    final op = entry['output_path'] as String? ?? '';
    _c.removeHistoryEntry(timestamp: ts, outputPath: op);
  }

  void _onClear() {
    AppModal.show(
      context,
      title: 'Clear history?',
      body: Text(
        'All compression history will be removed. This cannot be undone.',
        style: AppTheme.of(context).typography.secondary,
      ),
      actions: [
        const CancelButton(),
        AppButton(
          label: 'Clear history',
          variant: ButtonVariant.destructive,
          icon: 'trash-2',
          onPressed: () {
            Navigator.of(context).pop();
            _c.clearHistory();
            ToastHost.of(context).info(
              'History cleared',
              'All compression history has been removed.',
            );
          },
        ),
      ],
    );
  }
}

class _ActionsMenu extends StatelessWidget {
  const _ActionsMenu({
    required this.onOpenLocation,
    required this.onCopyPath,
    required this.onRemove,
  });

  final VoidCallback onOpenLocation;
  final VoidCallback onCopyPath;
  final VoidCallback onRemove;

  @override
  Widget build(BuildContext context) {
    final palette = AppTheme.of(context).palette;
    return PopupMenuButton<String>(
      icon: Icon(Icons.more_horiz, size: 18, color: palette.textMuted),
      color: palette.card,
      shape: RoundedRectangleBorder(
        borderRadius: AppRadius.mdAll,
        side: BorderSide(color: palette.border),
      ),
      onSelected: (value) {
        switch (value) {
          case 'open':
            onOpenLocation();
          case 'copy':
            onCopyPath();
          case 'remove':
            onRemove();
        }
      },
      itemBuilder: (_) => [
        PopupMenuItem(
          value: 'open',
          child: Row(
            children: [
              AppIcon('folder-open', size: 14, color: palette.textSecondary),
              const SizedBox(width: 8),
              const Expanded(
                child: Text('Open file location', overflow: TextOverflow.ellipsis),
              ),
            ],
          ),
        ),
        PopupMenuItem(
          value: 'copy',
          child: Row(
            children: [
              AppIcon('copy', size: 14, color: palette.textSecondary),
              const SizedBox(width: 8),
              const Expanded(
                child: Text('Copy output path', overflow: TextOverflow.ellipsis),
              ),
            ],
          ),
        ),
        const PopupMenuDivider(),
        PopupMenuItem(
          value: 'remove',
          child: Row(
            children: [
              AppIcon('trash-2', size: 14, color: palette.danger),
              const SizedBox(width: 8),
              Expanded(
                child: Text(
                  'Remove from history',
                  overflow: TextOverflow.ellipsis,
                  style: TextStyle(color: palette.danger),
                ),
              ),
            ],
          ),
        ),
      ],
    );
  }
}
